src/resources/log-message.ts

```diff
@@ -0,0 +1,70 @@
+import "core-js/modules/web.url.can-parse";
+import { fromError } from "stacktrace-js";
+import { UAParser } from "ua-parser-js";
+
+// URL paths to remove from filenames and max stack trace lines for brevity
+const REMOVAL_PATHS =
+  /^\/(?:home-assistant\/frontend\/[^/]+|unknown|\/{2}\.)\//;
+const MAX_STACK_FRAMES = 10;
+
+export const createLogMessage = async (
+  error: unknown,
+  intro?: string,
+  messageFallback?: string,
+  stackFallback?: string
+) => {
+  const lines: (string | undefined)[] = [];
+  // Append the originating browser/OS to any intro for easier identification
+  if (intro) {
+    const parser = new UAParser();
+    const {
+      name: browserName = "unknown browser",
+      version: browserVersion = "",
+    } = parser.getBrowser();
+    const { name: osName = "unknown OS", version: osVersion = "" } =
+      parser.getOS();
+    const browser = `${browserName} ${browserVersion}`.trimEnd();
+    const os = `${osName} ${osVersion}`.trimEnd();
+    lines.push(`${intro} from ${browser} on ${os}`);
+  }
+  // In most cases, an Error instance will be thrown, which can have many details to log:
+  // - a standard string coercion to "ErrorType: Message"
+  // - a stack added by browsers (which must be converted to original source)
+  // - an optional cause chain
+  // - a possible list of aggregated errors
+  if (error instanceof Error) {
+    lines.push(error.toString() || messageFallback);
+    const stackLines = (await fromError(error))
+      .slice(0, MAX_STACK_FRAMES)
+      .map((frame) => {
+        frame.fileName ??= "";
+        // @ts-expect-error canParse not in DOM library yet
+        if (URL.canParse(frame.fileName)) {
+          frame.fileName = new URL(frame.fileName).pathname;
+        }
```

Comment:
You know this would be the same in a try/catch right?

```js
try {
          frame.fileName = new URL(frame.fileName).pathname;
} catch (_e) { 
	// ignore
}

Reply:
Yeah but this is faster and shorter 😉.  It's also already in Firefox and Safari 17.